staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher_test.go

```diff
@@ -424,3 +437,85 @@ func testSetupWithEtcdServer(t *testing.T, opts ...setupOption) (context.Context
 
 	return ctx, cacher, server, terminate
 }
+
+func testSetupWithFakedStorage(t *testing.T, opts ...setupOption) (storage.Interface, func()) {
+	var fakeStorage *dummyStorage
+	var watcher *watch.FakeWatcher
+	var cacher *Cacher
+	var err error
+
+	setupOpts := setupOptions{}
+	opts = append([]setupOption{withDefaults}, opts...)
+	for _, opt := range opts {
+		opt(&setupOpts)
+	}
+
+	createCounter := 1
+	fakeStorage = &dummyStorage{}
+	fakeStorage.getListFn = func(_ context.Context, _ string, _ storage.ListOptions, listObj runtime.Object) error {
+		podList := listObj.(*example.PodList)
+		podList.ListMeta = metav1.ListMeta{ResourceVersion: fmt.Sprintf("%d", createCounter)}
+		return nil
+	}
+	podsForKey := map[string][]*example.Pod{}
+	fakeStorage.createFn = func(_ context.Context, _ string, obj, _ runtime.Object, _ uint64) error {
+		// since we are using a dummy storage
+		// simulate progress on the RV on each addition
+		createCounter++
+		pod := obj.(*example.Pod)
+		pod.ResourceVersion = fmt.Sprintf("%d", createCounter)
+
+		// collect all pods for the given key
+		// added so far for a potential watch request
+		key := fmt.Sprintf("/pods/%s", pod.Namespace)
+		pods := podsForKey[key]
+		pods = append(pods, pod)
+		podsForKey[key] = pods
```

Comment:
This is all super hack (what if e.g. I will decide to add a case to watch an individual pod now)?

Let's state explicitly what you want to achieve here - unless I'm missing something all you need is to ensure that after you create an object you just want to ensure that it actually is in the watchcache.

So instead of this, I would do something much simpler which is;
- use a real etcd underneath [that way we really will be testing what there really is underneath in prod]
- add a simple test wrapper around Cacher that just reimplments the Create method like:

```
type testCacher struct {
  Cacher
}

func (c *testCacher) Create(...) error {
  err := c.Cacher.Create(...)
  if err != nil {
    return err
  }
  return wait.Poll(..., func() {
   var currentObj
    err := t.cacher.Get(ctx, GetOptions{ResourceVersion: "0"}, &currentObj)
    if !api.Semantic.DeepEqual(currentObj, out) {
      return false
     }
     return true
  }
}
```

This avoids to reimplement a bunch of logic in a custom and not always working ways...

Reply:
since we are testing cacher which is a distinct implementation of the storage layer we should not use the real etcd underneath otherwise we cannot be sure which implementation we use

Reply:
> This is all super hack (what if e.g. I will decide to add a case to watch an individual pod now)?

as always, you would have to adjust the test to do what you want, i don't see there is a value in making this code future-proof

Reply:
ok, updated, ptal.